builder/dockerfile/dispatchers.go

```diff
@@ -383,64 +367,69 @@ func run(b *Builder, args []string, attributes map[string]bool, original string)
-	hit, err := b.probeCache()
-	if err != nil {
+	// TODO: this was previously in b.create(), why is it necessary?
+	b.runConfig.Image = b.image
+
+	runConfigForCacheProbe := copyRunConfig(b.runConfig, withCmd(saveCmd))
+	hit, err := b.probeCache(b.image, runConfigForCacheProbe)
+	switch {
+	case err != nil:
 		return err
-	}
-	if hit {
+	case hit:
 		return nil
 	}
 
-	// set Cmd manually, this is special case only for Dockerfiles
-	b.runConfig.Cmd = config.Cmd
-	// set build-time environment for 'run'.
-	b.runConfig.Env = append(b.runConfig.Env, cmdBuildEnv...)
+	runConfig := copyRunConfig(b.runConfig,
+		withCmd(cmdFromArgs),
+		withEnv(append(b.runConfig.Env, buildArgs...)))
+
 	// set config as already being escaped, this prevents double escaping on windows
-	b.runConfig.ArgsEscaped = true
+	// TODO: does this need to be set on the global b.runConfig as well?
+	runConfig.ArgsEscaped = true
 
-	logrus.Debugf("[BUILDER] Command to be executed: %v", b.runConfig.Cmd)
+	logrus.Debugf("[BUILDER] Command to be executed: %v", runConfig.Cmd)
 
-	cID, err := b.create()
+	cID, err := b.create(runConfig)
 	if err != nil {
 		return err
 	}
-
-	if err := b.run(cID); err != nil {
+	if err := b.run(cID, runConfig.Cmd); err != nil {
 		return err
 	}
 
-	// FIXME: this is duplicated with the defer above in this function (i think?)
-	// revert to original config environment and set the command string to
-	// have the build-time env vars in it (if any) so that future cache look-ups
-	// properly match it.
-	b.runConfig.Env = env
-
-	// remove builtinAllowedBuildArgs (see: builder.go)  from the saveCmd
-	// these args are transparent so resulting image should be the same regardless of the value
-	if len(cmdBuildEnv) > 0 {
-		saveCmd = config.Cmd
-		tmpBuildEnv := make([]string, len(cmdBuildEnv))
-		copy(tmpBuildEnv, cmdBuildEnv)
-		for i, env := range tmpBuildEnv {
+	// remove any unreferenced built-in args from the environment variables
+	// before committing. These args are transparent so resulting image should be
+	// the same regardless of the value
+	if len(buildArgs) > 0 {
+		for i, env := range buildArgs {
 			key := strings.SplitN(env, "=", 2)[0]
 			if b.buildArgs.IsUnreferencedBuiltin(key) {
-				tmpBuildEnv = append(tmpBuildEnv[:i], tmpBuildEnv[i+1:]...)
+				buildArgs = append(buildArgs[:i], buildArgs[i+1:]...)
 			}
 		}
-		sort.Strings(tmpBuildEnv)
-		tmpEnv := append([]string{fmt.Sprintf("|%d", len(tmpBuildEnv))}, tmpBuildEnv...)
-		saveCmd = strslice.StrSlice(append(tmpEnv, saveCmd...))
+		saveCmd = prependEnvOnCmd(buildArgs, cmdFromArgs)
 	}
-	b.runConfig.Cmd = saveCmd
-	return b.commit(cID, cmd, "run")
+
+	// FIXME: Previously this was relying on b.runConfig being a reference to
+	// the container.Config that was created, and is about to be commited.
+	// daemon.Commit() does a daemon.getContainer() to build the history, and
+	// this modifies the config after the container has been run, changing it
+	// to the updated Cmd
+	// saveCmd
```

Comment:
@tonistiigi this is a big surprise to me. I wonder if it is to you as well.

I guess this has been the case for a while, but nothing relied on it until #31584. It looks like there was lots of discussion about design and docs in that issue, but not much about the implementation.

This was relying of the fact ContainerCreate used the reference directly, so it was possible for the builder to modify `runConfig` and it would change the config of the already created container! 

I'm trying to find a way to implement this without this hack. I think `Commit()` will need to change to accept more args? Maybe it should accept the history?